kythe/typescript/indexer.ts

```diff
@@ -31,12 +31,44 @@ enum RefType {
   READ_WRITE
 }
 
+/**
+ * Various options that are required in order to perform indexing.
+ */
+export interface IndexingOptions {
+  /** A VName for the entire compilation, containing e.g. corpus name. */
+  compilationUnit: VName;
+
+  /** A map of file path to path-specific VName. */
+  pathVNames: Map<string, VName>;
```

Comment:
What do you of breaking this into two components:

`compilationUnit`: has VName, pathVNames, and srcfiles (the latter  are currently the paths passed to `index`)

`IndexingOptions` (or `IndexingEnvironment`): has the `emit` and  `plugins`

So what about `program` and `readFile` .... can the indexer itself create the program object? Pass in more stuff in `options` if needed for that. 

`readFile`. Hmm. Maybe let's make a full/complete `CompilationUnit` object? Put sources, the vname map, the root vname, and the readFile function on that?

Reply:
Good suggestion. I kept readFile() in IndexingOptions though. For me CompilationUnit is "dynamic" part, that changes from indexing to indexing, while IndexingOptions normally static and depends on environment and command line arguments. I think even CompilerOptions while currently coming from compilation units, eventually should become static and we should apply same CompilerOptions for all units. 

Note that this made `index()` method signature to diverge more so it will require more work on merging to google3. Though the merging is still relatively trivial.

Reply:
I know the implementation of readFile has to depend on the compilation unit, so it seems a little odd that it's passed separately from the CU, and just once at creation. But conceptually it doesn't belong on the CU, so I think this works.